### ISSUE TITLE
Opentelemetry for database/sql

### ISSUE THREAD

**Issue by wingyplus** [open-telemetry/opentelemetry-go-contrib#5]
It's good to see support for database/sql driver. We've some discussion here https://github.com/open-telemetry/opentelemetry-go/pull/490.

**Comment by andrewhsu** [comment]
FYI initial attempt that looks like a different approach would be more desirable https://github.com/open-telemetry/opentelemetry-go-contrib/pull/32

**Comment by seslattery** [comment]
I have an initial proof of concept I implemented at https://github.com/seslattery/otelsql.  If this seems like a useful direction to be explored more, I'd be happy to open a PR and work on polishing it up some more.

**Comment by Aneurysm9** [comment]
@seslattery please do open a PR.  I'd like to see `Register()` use functional options for the tracer and attributes, to be more in line with the SDK and other instrumentations.  I'm also not sure about using `runtime.Caller()` to derive span names from callers rather than just putting static strings at the call site.  Overall, though, this is a good start.

**Comment by MrAlias** [comment]
Prior work in OpenCensus might be useful to look at: https://github.com/opencensus-integrations/ocsql

I would not copy this, but might be interesting to look at.

**Comment by MrAlias** [comment]
Also: https://github.com/DataDog/dd-trace-go/tree/v1.24.1/contrib/database/sql

**Comment by MrAlias** [comment]
Prior attempt within this repo: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/32

**Comment by seanschade** [comment]
@seslattery are you still planning on submitting a PR?

> I have an initial proof of concept I implemented at https://github.com/seslattery/otelsql. If this seems like a useful direction to be explored more, I'd be happy to open a PR and work on polishing it up some more.

@MrAlias what are your thoughts on using something like https://github.com/ngrok/sqlmw as an interceptor to wrap drivers?

I think an approach similar to https://github.com/seslattery/otelsql is more desirable than https://github.com/open-telemetry/opentelemetry-go-contrib/pull/32. 

If we would like to forego the wrapper approach then we have both [OpenTracing](https://github.com/DataDog/dd-trace-go/tree/v1.24.1/contrib/database/sql) and [OpenCensus](https://github.com/opencensus-integrations/ocsql) examples to draw from. 

It would be nice if we could enumerate what we like about each, and what this should look like for OpenTelemetry.

**Comment by seslattery** [comment]
@seanschade Thanks for the ping!  Sorry for the delay, this fell off my radar.  I cleaned up the code a little bit and added an example.  PR is here: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/320.  Notably my implementation still doesn't have any tests.

**Comment by derekperkins** [comment]
Looking at that PR, it would be nice to be able to opt into specific spans. The SQL driver can be very chatty, so allowing opt-in like the ocsql package would be nice.
https://github.com/opencensus-integrations/ocsql/blob/master/options.go#L12-L64

**Comment by MrAlias** [comment]
> @MrAlias what are your thoughts on using something like https://github.com/ngrok/sqlmw as an interceptor to wrap drivers?

I agree with @derekperkins on the approach. I think an approach similar to what OpenCensus did is preferred to the approach of using the `sqlmw` approach.

> It would be nice if we could enumerate what we like about each, and what this should look like for OpenTelemetry.

This sounds like a great idea :+1: . I think including something like this in this issue would help future developers understand the eventual choice made and what out strategy was.

**Comment by faceair** [comment]
I think newrelic's sqldriver injector is simpler and more elegant.

https://github.com/newrelic/go-agent/blob/v3.9.0/v3/newrelic/sql_driver.go
https://github.com/newrelic/go-agent/blob/v3.9.0/v3/newrelic/sql_driver_optional_methods.go

**Comment by j2gg0s** [comment]
Is there any progress about this issue?
If we need an approache like opencensus, how about https://github.com/j2gg0s/otsql , used in our other project.

**Comment by kevindavee** [comment]
hi, any progress about this issue?

**Comment by MrAlias** [comment]
Is there someone is this issue that could head up this effort? This issue has not seen much focus from the maintainer/approver development team of this project as it is not a part of the things needed for our GA release (there are quite a lot of [issues](https://github.com/open-telemetry/opentelemetry-go/projects/1) prioritized for that effort that need to take precedence).

That said, it seems like there is a fair amount of community interest here. It would be ideal if someone can take the initiative and start work on what @seanschade laid out as a first step to a solutions design:

> I think an approach similar to https://github.com/seslattery/otelsql is more desirable than #32.
> 
> If we would like to forego the wrapper approach then we have both OpenTracing and OpenCensus examples to draw from.
> 
> It would be nice if we could enumerate what we like about each, and what this should look like for OpenTelemetry.

From the outcome of that analysis the implementation of a solution would need to be iterative. We've discussed a few times in our SIG meetings that the initial solution here does not need to solve all the combinatorial possibilities (like #32) but provided more of an MVP to build from.

**Comment by shousper** [comment]
+1 for @j2gg0s's implementation. `otelsql` only has a tracing implementation.

Metrics are far more important to us than tracing as they power alerting, dashboards and automation.

**Comment by seanschade** [comment]
> +1 for @j2gg0s's implementation. `otelsql` only has a tracing implementation.
> 
> Metrics are far more important to us than tracing as they power alerting, dashboards and automation.

This solution would meet my needs. Could @j2gg0s solution be the MVP, and would it move into the contrib repo?

**Comment by dackroyd** [comment]
Are alternative designs being considered? `ocsql`, `otelsql` (which is based on that), and `instrumentedsql` (OpenTracing option, e.g. https://github.com/luna-duclos/instrumentedsql) are all _very verbose_ in the number of spans that are created (as @derekperkins pointed out), most of which are of low value by themselves (e.g. 'rows next'). While it is possible to disable creation of spans for specific things, instrumenting this way may not be the most effective way to make things observable?

Using an observability tool like Honeycomb/Lightstep, you want as many details available on the one span as possible as that allows those things to be properly queried, and expose where other tags differ on those kinds of spans. The way I imagine that would be something like:

```
name: "SELECT"
tags:
  # Use proper query text per OTel spec
  db.statement: "SELECT ... FROM ... WHERE ..."
  # Statement duration before iterating over rows
  db.statement_duration_ms: 38.2
  # Need to be considered carefully, given args may include sensitive data like PII, or could have large values
  db.arg.foo: "bar"
  db.arg.fizz: "buzz"
  # Could be accumulated as the calling code iterates over the rows, added on `rows.Close()`
  db.rows_returned: 23
  ...
  # Spec tags that apply to all DB spans
  db.system: "postgresql"
  db.user: "username"
  db.name: "Customers"
  ...
```

I'd also be looking to hook into these details to aggregate things like number of statements executed, rows returned, total duration to put into the root span for the service e.g. the HTTP request that needs those queries executed.

**Comment by MrAlias** [comment]
From the SIG meeting today:

> @Aneurysm9 mentioned that possibly Honeycomb might be able to contribute their DB instrumentation so it could be converted and used as OTel instrumentation. @Aneurysm9 is going to follow up with @lizthegrey.

Also,

> As there are multiple approaches to provide this instrumentation we should likely add multiple instrumentation packages for users to try. Ideally we will ultimately provide a single well vetted and supported instrumentation library for the `db` package, but until then we should add add an experimental directory and add all 3 (possibly 2) solutions listed here to that directory so they can be iterated on and users can evaluate them.

**Comment by chainlink** [comment]
Any update on the SIG discussion?

**Comment by XSAM** [comment]
I'd like to work on this by implementing an MVP (probably not include metrics first)

**Comment by derekperkins** [comment]
I've been watching this issue, but I wasn't aware that @XSAM had indeed built a version https://github.com/XSAM/otelsql. In his PR, @Aneurysm9 said this about landing official support https://github.com/open-telemetry/opentelemetry-go-contrib/pull/505#issuecomment-800452510

> We've discussed this a couple times at the SIG meeting and the consensus seems to be that we are not ready to land a db/sql instrumentation in the contrib repo at this time. We would like to see broader usage and community consensus on an approach before we commit to the level of support that would be required of a package in contrib.

**Comment by lizthegrey** [comment]
cc @bdarfler @adamopenweb

**Comment by derekperkins** [comment]
@MrAlias any feedback on this?

**Comment by derekperkins** [comment]
For anyone subscribed to or who finds this issue, this fork https://github.com/XSAM/otelsql is active and probably your best bet for the time being.

**Comment by tonglil** [comment]
Has anyone done a comparison between https://github.com/XSAM/otelsql and https://github.com/uptrace/opentelemetry-go-extra/ ?
